Do not ping MySQL on connection checkout

kirs opened the pull request:
Now on every single connection checkout we ping MySQL with
`mysql_ping`. This is nothing for an app with a couple app servers, but in case of Shopify we have over 500 servers, each running 4 containers with 12 unicorn processes. It means that on the first request after the deploy, we checkout 500 * 4 * 12 = **24000** connections. This creates a peak of 24K QPS on the master MySQL database and makes our data ops sad.

**Why is this bad**? Because these pings are not free. On every request we (usually) checkout two connections: 1) the shard, and 2) the master shard. This is two network roundtrips of roughly 1ms each. In addition, these calls are not free on the database either. Each call takes 100μs, whereas a small, indexed query takes 500μs. Given that we have such a staggering number of them, especially on the master shard which is checked out on connections to all shards, this adds up to an absurd amount of CPU time spent in Rails pinging connections and on the database processing those pings.

**Why is the ping there**? To health-check the connection and reconnect if it's not healthy. This is problematic though, because it's racy. If you perform the ping, and reconnect, by the time you hit the first query your connection might have been lost. For the majority of production, this is a no-op ping. It's legitimate in cases where the connection has gone stale (even that is rare). If we blindly remove the ping, in those cases we'd lose the entire request. This is not unreasonable, but this patch will ensure that even requests from stale or closed connections will succeed by retrying the first query after a connection checkout.

**This is not only the Shopify problem**. Same issue have been described by [Percona](https://www.percona.com/blog/2010/05/05/checking-for-a-live-database-connection-considered-harmful/) and [by Groupon on their Rails patch](https://engineering.groupon.com/2013/rails/removing-the-ping-sting-from-rails-db-connections/). As outlined in Percona blog post, 73% of the server’s load was consumed by checking to see if the connection is still alive.

**Summary:** while the code behind my implementation is not ideal, I think that many large-scale Rails apps such as Basecamp will benefit from the option to skip MySQL ping. I'm open to the discussion how we can make it an optional thing.

All credits for discovering the issue and the patch belong to @Sirupsen and @rafaelfranca.

@tenderlove @arthurnn @sgrif @jeremy

matthewd commented:
👍🏻 on getting this in. I went looking for a prior PR, but couldn't find one. Did no-one ever submit Groupon's version?

... but let's do this properly, rather than upstream a monkey-patch. 😢 

* It should be in the base adapter, not MySQL
* Why the double-underscore?
* `verify!` should continue to do what it says on the tin (possibly deprecated, if we no longer have an internal use for such direct action, though it still sounds useful to me -- e.g. that test); instead, the connection pool should switch to a softer `you_are_clean_and_have_no_state_and_therefore_may_reconnect_on_query_failure!`
* I likewise see no reason to make `active?` force a reconnect
* Tests plz. It seems unlikely this drops in quite so well as to be just properly covered by the existing tests

BTW, if you're up for a larger "defer things until the first real query" adventure, https://github.com/rails/rails/issues/17937 could save several round-trips in some cases. I haven't closely thought through whether it directly overlaps with what we're doing here, but it feels like it's in the same territory, at least.

---

> If we blindly remove the ping, in those cases we'd lose the entire request. This is not unreasonable

I know you're not recommending that, but for the record: strong disagree. Losing a request because the database fails in the middle is the price of doing business. Losing a request because the database server restarted 3 hours ago and this is the first time we've grabbed this connection from the pool since then, is entirely unreasonable.

sirupsen commented:
Thanks for bringing this upstream @kirs. Definitely agree with @matthewd that this shouldn't be specific to the MySQL driver upstream.

This patch came about when I noticed the amount of pings we did. When this patch was shipped for us, we _cut QPS in half_. I think this graph speaks for itself in terms of why this patch is worth the effort for large-scale production deployments:

![](https://cloud.githubusercontent.com/assets/97400/17228967/64952ffc-54e3-11e6-8fbd-3d5ae0e4ba2a.png)

I'd like to explain a couple of technical decisions as follow-ups to @matthewd's comment and great questions:

> Tests plz. It seems unlikely this drops in quite so well as to be just properly covered by the existing tests

We have tests in our own test-suite for the monkey-patch, that use [Toxiproxy](http://github.com/shopify/toxiproxy). Toxiproxy is not supported upstream, but as an aside, I think we should strongly consider bringing it in. @rafaelfranca and I have talked about this in the past. We end up finding Rails bugs, especially in the ActiveRecord connection pooling logic, in many of the major releases due to extensive resiliency coverage in our own test suite. If anyone is interested in working on this, I'd be happy to help.

> I likewise see no reason to make active? force a reconnect

I think that `active?` remains a useful deterministic way of checking whether the database is alive. If we don't reconnect, we can't do that, and we rely on this in checking whether it was an execution error (we can ping the connection) or a connection error (we can't). 

> Losing a request because the database server restarted 3 hours ago and this is the first time we've grabbed this connection from the pool since then, is entirely unreasonable.

Yeah, that makes sense for the general use-case. In our specific case, that wouldn't happen.

> Did no-one ever submit Groupon's version?

I originally looked at Groupon's version, but it has a couple of problems. It parses the errors in the raw, it looks at transactions (even though it doesn't need to), and there were another couple of problems with it. The patch we use internally is a major simplification of it.

Something that we also have internally that was _not_ submitted upstream is that we only retry within a reasonable time. We saw in production 25s time-out queries that were the first of a request being retried (and the DB struggling with them twice) because there was no time-out on the `@__verify_connection` check.

kirs committed:
Lazy connection verification

We used to verify connection by doing `SELECT 1` query for PG or `mysql_ping` call for MySQL.

This commit moves away from that because pings are not free
and this way on health checks is not the most reliable.
More details explained in the PR.

kirs commented:
@matthewd thanks for review! I updated the PR.

> I likewise see no reason to make `active?` force a reconnect

Calling `Mysql2::Client#ping` (that's what `active?` currently does) will force reconnect anyway.

kirs commented on activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:
```diff
@@ -107,6 +120,12 @@ def disconnect!
       private
 
+        def connected?
+          @connection && @connection.socket.present?
```
This is sketchy, but we'll be able to clean that up as soon as https://github.com/brianmario/mysql2/pull/796 get merged.

kirs replied:
@dylanahsmith does this look like what you're suggest?

kirs commented:
@matthewd mind giving another round of 👀 when you have a minute? Thanks!

arthurnn commented on activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:
```diff
@@ -79,16 +79,29 @@ def error_number(exception)
+        if @connection.ping
+          true
+        elsif @verify_connection
+          reconnect!
```
Not sure i like the implementation of `active?` method calling a `reconnect!`.

kirs replied:
I see the point here, but before my changes we called `connection.ping` in `active?` that would also reconnect if the connection is lost.

arthurnn commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -501,6 +510,21 @@ def combine_bind_parameters(
+        rescue exception_class
+          raise_unless_verifying_connection
+          reconnect!
+          retry
```
Should we have a limit on retries here?

kirs replied:
Sure

dylanahsmith replied:
This is limited to one retry, since it will only retry when `@verify_connection` is truthy and `raise_unless_verifying_connection` sets it to `@verify_connection = false`.

dylanahsmith replied:
Will this retry `INSERT` or `UPDATE` queries?  If so, the connection could get lost between when the query is committed and the response is received.  That means we could end up double inserting.

arthurnn commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -501,6 +510,21 @@ def combine_bind_parameters(
+        end
+
+        def raise_unless_verifying_connection
+          raise if !@verify_connection || active?
```
could we ever get a `@verify_connection` not defined warning in here? as the `defer_verify` method first defines that ivar.
Also, I am not 100% what this flag is for. Can you expand?

arthurnn reviewed (commented):
I think this is super valuable. 
We have some similar code at GitHub mysql adpter too, to reconnect.
Our implementation is a bit different tho.
We have a `query_with_retry` method, which is similar to `with_connection_verify`, but that always verify the connection, not based in a flag.
We also have some retry limit in that method.
And also that method does some dead locking retry on mysql, which is beyond the scope in here, but might be something we could think about when designing this hook, so folks can easily hook such functionality.

kirs commented:
> We have a query_with_retry method, which is similar to with_connection_verify, but that always verify the connection, not based in a flag.

Always verifying the connection means calling `active?` [(abstract_adapter.rb)][1] that will trigger `connection.ping` [(mysql_adapter.rb)][2]. Having that we'll still always have the PING queries on connection checkout. Are you sure that's how it works?

[1]: https://github.com/rails/rails/blob/3cb27c5/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L440

[2]: https://github.com/rails/rails/blob/d7b3334/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L90

pschambacher commented:
I would also be interested to see this go through and I'd like to add Zendesk to the list of companies affected by the ping issue 😄

kirs commented:
@pschambacher glad to hear that! I would love your feedback regarding https://github.com/rails/rails/issues/28200

dylanahsmith commented:
We had a problem at Shopify with a version of the patch like this where it would retry slow queries if they are the first query performed in handling a request.  This had the effect of doubling those slow queries which had their connection killed from the query taking too long.

We worked around that by timing how long the query took so that we could avoid retrying a query if it took more than a threshold amount of time.

However, I think the simpler solution would be to just check if the socket is closed without making any ping/query rather than deferring verification.

pschambacher commented:
This was just merged into the mysql2 gem: https://github.com/brianmario/mysql2/pull/796

kirs commented:
Thanks for letting me know! I'll update the PR soon.

On Jun 1, 2017 4:47 PM, "Pierre Schambacher" <notifications@github.com>
wrote:

> This was just merged into the mysql2 gem: brianmario/mysql2#796
> <https://github.com/brianmario/mysql2/pull/796>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/27651#issuecomment-305615333>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAf3q7Fvlto7ece4RGcUDm5aiS3fsK_Yks5r_yN4gaJpZM4LhMiu>
> .
>

padi commented:
@kirs given that you've written an alternative implementation for this issue https://github.com/rails/rails/pull/28200, either that PR or this PR should be closed.

I prefer the alternative since it seems like you've
1. stiłl addresses issues initially posted by @matthewd 
2. reduced surface area.

While we have brianmario/mysql2#796 already merged, looks like the `pg` gem did not chose to similar method: https://bitbucket.org/ged/ruby-pg/issues/129/reconnect-dropped-connections. tl;dr they'd rather let the application handle the reconnection and raise a unique exception instead.

kirs commented:
@dylanahsmith I've been looking into using `connection.closed?` from the recent version of mysql2 gem to implement this, but it looks like it doesn't catch the case when MySQL server closes idle connection:

```ruby
require 'mysql2'
client = Mysql2::Client.new(:host => "localhost", :username => "root")
client.query("set @@wait_timeout=1")
sleep 2
# connection is now closed by server
puts client.closed?
=> false
puts client.ping
=> false
puts client.closed?
=> true
```

We still need to ping connection after it has been closed by the server, otherwise the socket is detected as "open". WDYT?

dylanahsmith commented:
@kirs you are right that it wouldn't detect cases where the server closes the connection.  That is something that prevented us from moving from the lazy verification approach in our monkey patch.  Now, we just retry any query on a connection error, made outside of a transaction and didn't timeout.

P9GIT commented:
@kirs @dylanahsmith etc.
We get hit by this as well - just we use PG instead. It's painful to see the amount of wasted resources :o(

It seems this issue of connection pings is still unresolved and each organization is doing their own thing to amend it.
It's a shame that great solutions are not added to AR or at least shared.

Thanks for all the input on this problem.

Thomas.

kirs commented:
@P9GIT we came to realize that while pingless strategy reduced a lot of load from MySQL for free, it also created some occasional edge cases when a request would run into a closed connection.

Running a multiplexing proxy in front of the DB (for us it's ProxySQL, for PG I'd guess it would be PGBouncer) helped us to remove this patch altogether and use the Rails way of sending PING packets on every request - at the end, those pings are handled by the proxy and won't create any extra load on the DB itself.

That would be my recommendation for anyone who sees load on the database coming either from a high number of connections or from ping packets. I'm not planning to continue to work on this PR so I'm going to close it.

P9GIT commented:
@kirs 

Thanks for your response.

Unfortunately in case of PG it's not just a ping package but AR does a "SELECT 1"-request which makes its way to the PG server through our PGBouncer.

Another observation I've made is that using "ActiveRecord::Base.connection.quote" to quote parameters (no ORM, we use direct SQL functions for most of our data) a connection is checked out of the pool which again triggers a "SELECT 1" ping. 
So we have loads of these :( 

Any pointers would be greatly appreciated!

Thomas.

kirs commented:
Isn’t there a way to make PGBouncer cache/mock results of SELECT 1?

On Mon, 10 Feb 2020 at 16:59, P9GIT <notifications@github.com> wrote:

> @kirs <https://github.com/kirs>
>
> Thanks for your response.
>
> Unfortunately in case of PG it's not just a ping package but AR does a
> "SELECT 1"-request which makes its way to the PG server through our
> PGBouncer.
>
> Another observation I've made is that using
> "ActiveRecord::Base.connection.quote" to quote parameters (no ORM, we use
> direct SQL functions for most of our data) a connection is checked out of
> the pool which again triggers a "SELECT 1" ping.
> So we have loads of these :(
>
> Any pointers would be greatly appreciated!
>
> Thomas.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/27651?email_source=notifications&email_token=AAD7PK2A4KFVGULHQWFB3VDRCF2U7A5CNFSM4C4EZCXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELJBSSI#issuecomment-584194377>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7PK2CMS2TWUQCITR2XTLRCF2U7ANCNFSM4C4EZCXA>
> .
>

simi commented:
Maybe something from libpq can be used instead.

https://www.postgresql.org/docs/current/libpq-status.html

P9GIT commented:
@kirs 
I'm afraid not. PGBouncer is only a connection pool cache. It has very limited features.

@simi 
Yes, I have subsequently been looking at the libpq "PQping" feature which it seems does the thing same as "SELECT 1" it seems.
This is exposed through "PG::Connection.ping" in ruby-pg. 
Maybe this is the answer. I'll give a shot!

Thanks!

matthewd commented:
@P9GIT  `PQstatus` doesn't attempt to talk to the remote server, and `PQping` doesn't use the existing connection; neither are suitable here.

simi commented:
@matthewd you're right. I have seen that pg go wrapper uses empty query (https://github.com/lib/pq/blob/9eb3fc897d6fd97dd4aad3d0404b54e2f7cc56be/notify.go#L239) aka `ActiveRecord::Base.connection.raw_connection.exec("")`. It talks to remote server. I have not time to check if this reaches the remote server thru pgbouncer as well.

P9GIT commented:
@matthewd @simi 
Aha, I see. I'll try out the empty query technique with our PGBouncer setup.

P9GIT commented:
@simi 

Alas, the empty string query does exactly the same as "SELECT 1". It ends up on the SQL server.
